front: fix invalid category import

Synar opened the pull request:
Close [#13075](https://github.com/OpenRailAssociation/osrd/issues/13075)

Also fix related type issues, see first 2 commits titles and descriptions.

Synar committed:
front: revert TrainMainCategoryDict to record

TrainMainCategoryDict was created as a passthrough record over an enum
or a list for the sole purpose of enforcing at compile time that the list
of categories used in the front is the exact same as the one
provided by the open api type (which no longer exists at runtime).

Using a map does not provide this same type safety.
Additionally, this is a passthrough dict, there is no reason to
use get on it.

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
front: fix checkCategory input type in train import

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
front: add unrecognized category as labels in train import

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

clarani commented on front/src/modules/rollingStock/consts.ts:
```diff
@@ -11,17 +11,19 @@ export const ComfortLevels: Record<Comfort, Comfort> = {
-]);
+// Note that this check works for records, but not maps or enums, as they check that all keys are of the proper type,
+// but not that all possible keys are present.
+export const TrainMainCategoryDict: Record<TrainMainCategory, TrainMainCategory> = {
```
Why not use an array ?

Synar replied:
See the comment and the commit description. An array doesn't need to contain all possible values of its type to be valid, so we could have strictly less categories included than in the openapi type. 

Sidenote : I think the most sensible data structure would be an enum here, but then to ensure perfect coherence with the openapi type we need an additional, unused type, which is a bit weird and complex.

We also could allow the type to differ from the open api type and assume we'll realize mistakes through usage, but I would rather enforce open api coherence if it's not too costly.

clarani replied:
Ok thanks, I got the nuance :)

clarani commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTimetableItem/ImportTimetableItemTrainsList.tsx:
```diff
@@ -88,15 +88,39 @@ const ImportTimetableItemTrainsList = ({
       pacedTrainsJsonData: pacedTrainsFromJsonData.map((pacedTrain) => ({
         ...pacedTrain,
         category: checkCategory(pacedTrain.category),
+        labels: buildLabels(pacedTrain.labels, pacedTrain.category),
```
Hum `pacedTrain.Category` can be a string here ? With the types, I thought it was only `TrainCategory | null | undefined`

Synar replied:
Yes, the types are lying a bit (and fixing it would be nice). We've casted the imported json to a type that doesn't allow strings earlier in the code, but we actually allow a fallback for jsons that still use a string here, for retro compatibility.

See this comment :

     // This condition is added for train imports that still use the old format: `category: string`

clarani replied:
Can you add a TODO to correct the types please ?

Synar replied:
Sure!

Synar replied:
Actually I just fixed them instead, the type change didn't have the snowball effect I thought it would have ^^

emersion commented on front/src/modules/rollingStock/consts.ts:
```diff
@@ -11,17 +11,19 @@ export const ComfortLevels: Record<Comfort, Comfort> = {
-]);
+// Note that this check works for records, but not maps or enums, as they check that all keys are of the proper type,
+// but not that all possible keys are present.
+export const TrainMainCategoryDict: Record<TrainMainCategory, TrainMainCategory> = {
```
I don't think using a `Record` is a good idea here. Records don't retain key types, everything is converted to `string`. `Map` doesn't have this issue.

emersion replied:
Oh my, I just say the rest of the description. I'm not in love with this approach, but I don't see a better way.

Synar replied:
One thing we can do is use a different data structure in practice from the one used to validate api type coherence, either by deriving the practical data structure from the record, or by using a different data structure from the get go and deriving from it a type used only for api type validation. There's added complexity either way though.

Synar committed:
front: accurately type TrainScheduleFromJson

Train schedules parsed from json can use string categories
(instead of structured categories in the form of TrainCategory objects)
before being processed into fully valid train schedules.

Reflect this in a new type.

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>
